File: app/src/main/java/com/zeapo/pwdstore/pwgen/RandomPasswordGenerator.java
package com.zeapo.pwdstore.pwgen;

class RandomPasswordGenerator {

    /**
     * Generates a completely random password.
     *
     * @param size    length of password to generate
     * @param pwFlags flag field where set bits indicate conditions the
     *                generated password must meet
     *                <table summary ="bits of flag field">
     *                <tr><td>Bit</td><td>Condition</td></tr>
     *                <tr><td>0</td><td>include at least one number</td></tr>
     *                <tr><td>1</td><td>include at least one uppercase letter</td></tr>
     *                <tr><td>2</td><td>include at least one symbol</td></tr>
     *                <tr><td>3</td><td>don't include ambiguous characters</td></tr>
     *                <tr><td>4</td><td>don't include vowels</td></tr>
     *                </table>
     * @return the generated password
     */
    public static String rand(int size, int pwFlags) {
        String password;
        char cha;
        int i, featureFlags, num;
        String val;

        String bank = "";
        if ((pwFlags & PasswordGenerator.DIGITS) > 0) {
            bank += PasswordGenerator.DIGITS_STR;
        }
        if ((pwFlags & PasswordGenerator.UPPERS) > 0) {
            bank += PasswordGenerator.UPPERS_STR;
        }
        bank += PasswordGenerator.LOWERS_STR;
        if ((pwFlags & PasswordGenerator.SYMBOLS) > 0) {
            bank += PasswordGenerator.SYMBOLS_STR;
        }
        do {
            password = "";
            featureFlags = pwFlags;
            i = 0;
            while (i < size) {
                num = RandomNumberGenerator.number(bank.length());
                cha = bank.toCharArray()[num];
                val = String.valueOf(cha);
                if ((pwFlags & PasswordGenerator.AMBIGUOUS) > 0
                        && PasswordGenerator.AMBIGUOUS_STR.contains(val)) {
                    continue;
                }
                if ((pwFlags & PasswordGenerator.NO_VOWELS) > 0
                        && PasswordGenerator.VOWELS_STR.contains(val)) {
                    continue;
                }
                password += val;
                i++;
                if (PasswordGenerator.DIGITS_STR.contains(val)) {
                    featureFlags &= ~PasswordGenerator.DIGITS;
                }
                if (PasswordGenerator.UPPERS_STR.contains(val)) {
                    featureFlags &= ~PasswordGenerator.UPPERS;
                }
                if (PasswordGenerator.SYMBOLS_STR.contains(val)) {
                    featureFlags &= ~PasswordGenerator.SYMBOLS;
                }
            }
        } while ((featureFlags & (PasswordGenerator.UPPERS | PasswordGenerator.DIGITS | PasswordGenerator.SYMBOLS))
                > 0);
        return password;
    }
}

File: app/src/main/java/com/zeapo/pwdstore/pwgen/RandomPasswordGenerator.kt
/*
 * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
package com.zeapo.pwdstore.pwgen

import com.zeapo.pwdstore.utils.hasFlag

object RandomPasswordGenerator {

    /**
     * Generates a random password of length [targetLength], taking the following flags in [pwFlags]
     * into account, or fails to do so and returns null:
     *
     * - [PasswordGenerator.DIGITS]: If set, the password will contain at least one digit; if not
     *   set, the password will not contain any digits.
     * - [PasswordGenerator.UPPERS]: If set, the password will contain at least one uppercase
     *   letter; if not set, the password will not contain any uppercase letters.
     * - [PasswordGenerator.LOWERS]: If set, the password will contain at least one lowercase
     *   letter; if not set, the password will not contain any lowercase letters.
     * - [PasswordGenerator.SYMBOLS]: If set, the password will contain at least one symbol; if not
     *   set, the password will not contain any symbols.
     * - [PasswordGenerator.NO_AMBIGUOUS]: If set, the password will not contain any ambiguous
     *   characters.
     * - [PasswordGenerator.NO_VOWELS]: If set, the password will not contain any vowels.
     */
    fun generate(targetLength: Int, pwFlags: Int): String? {
        val bank = listOfNotNull(
            PasswordGenerator.DIGITS_STR.takeIf { pwFlags hasFlag PasswordGenerator.DIGITS },
            PasswordGenerator.UPPERS_STR.takeIf { pwFlags hasFlag PasswordGenerator.UPPERS },
            PasswordGenerator.LOWERS_STR.takeIf { pwFlags hasFlag PasswordGenerator.LOWERS },
            PasswordGenerator.SYMBOLS_STR.takeIf { pwFlags hasFlag PasswordGenerator.SYMBOLS }
        ).joinToString("")

        var password = ""
        while (password.length < targetLength) {
            val candidate = bank.secureRandomCharacter()
            if (pwFlags hasFlag PasswordGenerator.NO_AMBIGUOUS &&
                candidate in PasswordGenerator.AMBIGUOUS_STR) {
                continue
            }
            password += candidate
        }
        return password.takeIf { PasswordGenerator.isValidPassword(it, pwFlags) }
    }
}
